polkadot/node/core/approval-voting/src/lib.rs

```diff
@@ -788,6 +815,28 @@ struct State {
 	clock: Box<dyn Clock + Send + Sync>,
 	assignment_criteria: Box<dyn AssignmentCriteria + Send + Sync>,
 	spans: HashMap<Hash, jaeger::PerLeafSpan>,
+	// Per block, candidate records about how long we take until we gather enough
+	// assignments, this is relevant because it gives us a good idea about how many
+	// tranches we trigger and why.
+	per_block_assignments_gathering_times:
+		HashMap<BlockNumber, HashMap<(Hash, CandidateHash), AssignmentGatheringRecord>>,
```

Comment:
Why `(Hash, CandidateHash)` and not just  `CandidateHash` key for nested hash map ?

Reply:
I would also think about a sufficiently large LRU map might be simpler to use here.

Reply:
> Why (Hash, CandidateHash) and not just CandidateHash key for nested hash map ?

The assignments are unique per block per candidate, so if you got the same Candidate in two different blocks separate assignments messages are sent.

> I would also think about a sufficiently large LRU map might be simpler to use here.

Yes, that would be an option as well, although I did not like that finding the sufficiently large number is a matter of how many cores you've got available which is different on different chains.

Reply:
Never mind I had brain fog will switch to an LRU map.